src/Disks/DiskWebServer.cpp

```diff
@@ -0,0 +1,336 @@
+
+        if (!metadata.tables_data.count(table_uuid))
+            return false;
+
+        if (!metadata.tables_data[table_uuid].count(directory_name))
+            return false;
+
+        const auto & files = metadata.tables_data[table_uuid][directory_name];
+        auto file = files.find(File(file_name));
+        if (file == files.end())
+            return false;
+
+        file_info = *file;
+        return true;
+    }
+
+    return false;
+}
+
+
+bool DiskWebServer::exists(const String & path) const
+{
+    LOG_DEBUG(log, "Checking existence of file: {}", path);
+
+    File file;
+    return findFileInMetadata(path, file);
+}
+
+
+std::unique_ptr<ReadBufferFromFileBase> DiskWebServer::readFile(const String & path, size_t buf_size, size_t, size_t, size_t, MMappedFileCache *) const
+{
+    LOG_DEBUG(log, "Read from file by path: {}", path);
+
+    File file;
+    if (!findFileInMetadata(path, file))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "File {} not found", path);
+
+    RemoteMetadata meta(uri, fs::path(path).parent_path() / fs::path(path).filename());
+    meta.remote_fs_objects.emplace_back(std::make_pair(getFileName(path), file.size));
+
+    auto reader = std::make_unique<ReadBufferFromWebServer>(uri, meta, getContext(), settings->max_read_tries, buf_size);
+    return std::make_unique<SeekAvoidingReadBuffer>(std::move(reader), settings->min_bytes_for_seek);
+}
+
+
+std::unique_ptr<WriteBufferFromFileBase> DiskWebServer::writeFile(const String & path, size_t, WriteMode)
+{
+    if (path.ends_with("format_version.txt"))
+        return std::make_unique<WriteBufferFromNothing>();
+
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Disk {} is read-only", getName());
+}
+
+
+DiskDirectoryIteratorPtr DiskWebServer::iterateDirectory(const String & path)
+{
+    LOG_DEBUG(log, "Iterate directory: {}", path);
+    String uuid;
+
+    if (RE2::FullMatch(path, ".*/store/"))
+        return std::make_unique<DiskWebDirectoryIterator<RootDirectory>>(metadata.tables_data, path);
+
+    if (!RE2::Extract(path, EXTRACT_UUID_PATTERN, "\\1", &uuid))
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot extract uuid for: {}", path);
+
+    /// Do not throw if it is not a query, but disk load.
+    bool can_throw = CurrentThread::isInitialized() && CurrentThread::get().getQueryContext();
+
+    try
+    {
+        if (!metadata.tables_data.count(uuid))
+            metadata.initialize(uri, settings->files_prefix, uuid, getContext());
+    }
+    catch (const Poco::Exception &)
+    {
+        const auto message = getCurrentExceptionMessage(false);
+        if (can_throw)
+        {
+            throw Exception(ErrorCodes::NETWORK_ERROR, "Cannot load disk metadata. Error: {}", message);
+        }
+
+        LOG_TRACE(&Poco::Logger::get("DiskWeb"), "Cannot load disk metadata. Error: {}", message);
+        /// Empty iterator.
+        return std::make_unique<DiskWebDirectoryIterator<Directory>>(metadata.tables_data[""], path);
```

Comment:
Is there a reason to use `metadata.tables_data[""]` instead of just `{}`?

Reply:
Because directory iterator holds a reference to Directory struct.

Reply:
Ok, got it. I see this line is almost the same as line 286, not sure that it will be better, but we could write here just `uuid.clear()`. But it's ok to leave it as is.